File: DSA/SortingAlgorithms/Code/CountingSort.c
#include <stdio.h>
#include <stdlib.h>

// Counting Sort in C
// Non-comparison sort - counts occurrences of each element

void countingSort(int arr[], int n) {
    int max = arr[0];
    for (int i = 1; i < n; i++) {
        if (arr[i] > max)
            max = arr[i];
    }
    
    int *count = (int *)malloc((max + 1) * sizeof(int));
    for (int i = 0; i <= max; i++)
        count[i] = 0;
    
    for (int i = 0; i < n; i++)
        count[arr[i]]++;
    
    for (int i = 1; i <= max; i++)
        count[i] += count[i - 1];
    
    int *output = (int *)malloc(n * sizeof(int));
    for (int i = n - 1; i >= 0; i--) {
        output[count[arr[i]] - 1] = arr[i];
        count[arr[i]]--;
    }
    
    for (int i = 0; i < n; i++)
        arr[i] = output[i];
    
    free(count);
    free(output);
}

// Counting sort with range
void countingSortRange(int arr[], int n, int min, int max) {
    int range = max - min + 1;
    int *count = (int *)malloc(range * sizeof(int));
    
    for (int i = 0; i < range; i++)
        count[i] = 0;
    
    for (int i = 0; i < n; i++)
        count[arr[i] - min]++;
    
    int index = 0;
    for (int i = 0; i < range; i++) {
        for (int j = 0; j < count[i]; j++) {
            arr[index++] = i + min;
        }
    }
    
    free(count);
}

void displayArray(int arr[], int n) {
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

int main() {
    int arr[] = {64, 34, 25, 12, 22, 11, 90};
    int n = 7;
    
    printf("Original Array: ");
    displayArray(arr, n);
    
    countingSort(arr, n);
    
    printf("Sorted Array (Counting Sort): ");
    displayArray(arr, n);
    
    // Test with small range
    int arr2[] = {4, 2, 3, 4, 1, 2, 4};
    printf("\nSmall range array: ");
    displayArray(arr2, 7);
    countingSort(arr2, 7);
    printf("After Counting Sort: ");
    displayArray(arr2, 7);
    
    // Test with duplicates
    int arr3[] = {5, 2, 8, 2, 9, 1, 5, 5};
    printf("\nArray with duplicates: ");
    displayArray(arr3, 8);
    countingSort(arr3, 8);
    printf("After Counting Sort: ");
    displayArray(arr3, 8);
    
    // Test with negative numbers using range
    int arr4[] = {-2, -5, 3, 1, -2, 0, 3};
    printf("\nArray with negative numbers: ");
    displayArray(arr4, 7);
    countingSortRange(arr4, 7, -5, 3);
    printf("After Counting Sort (Range): ");
    displayArray(arr4, 7);
    
    printf("\nComplexity Analysis:\n");
    printf("Time Complexity: O(n + k) where k is range of input\n");
    printf("Space Complexity: O(k)\n");
    printf("Best Case: O(n + k)\n");
    printf("Worst Case: O(n + k)\n");
    printf("Average Case: O(n + k)\n");
    printf("Stable: Yes\n");
    printf("Note: Non-comparison, efficient for small ranges\n");
    
    return 0;
}

File: DSA/SortingAlgorithms/Code/CountingSort.py
# Counting Sort in Python
# Non-comparison sort - counts occurrences of each element

def counting_sort(arr):
    """Counting sort for non-negative integers"""
    if not arr:
        return arr
    
    max_val = max(arr)
    min_val = min(arr)
    range_val = max_val - min_val + 1
    
    count = [0] * range_val
    
    for num in arr:
        count[num - min_val] += 1
    
    for i in range(1, range_val):
        count[i] += count[i - 1]
    
    output = [0] * len(arr)
    for i in range(len(arr) - 1, -1, -1):
        output[count[arr[i] - min_val] - 1] = arr[i]
        count[arr[i] - min_val] -= 1
    
    return output

def counting_sort_range(arr, max_val):
    """Counting sort for range [0, max_val]"""
    count = [0] * (max_val + 1)
    
    for num in arr:
        count[num] += 1
    
    output = []
    for i in range(max_val + 1):
        output.extend([i] * count[i])
    
    return output

def counting_sort_by_digit(arr, exp):
    """Counting sort based on digit for radix sort"""
    n = len(arr)
    output = [0] * n
    count = [0] * 10
    
    for i in range(n):
        index = (arr[i] // exp) % 10
        count[index] += 1
    
    for i in range(1, 10):
        count[i] += count[i - 1]
    
    for i in range(n - 1, -1, -1):
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1
    
    return output

# Test 1: Basic counting sort
arr = [64, 34, 25, 12, 22, 11, 90]
print("Original Array:", arr)
result = counting_sort(arr.copy())
print("Sorted Array:", result)

# Test 2: Small range
arr2 = [4, 2, 3, 4, 1, 2, 4]
print("\nSmall range array:", arr2)
result2 = counting_sort_range(arr2.copy(), 4)
print("After Counting Sort:", result2)

# Test 3: Duplicates
arr3 = [5, 2, 8, 2, 9, 1, 5, 5]
print("\nArray with duplicates:", arr3)
result3 = counting_sort(arr3.copy())
print("After Counting Sort:", result3)

# Test 4: Already sorted
arr4 = [1, 2, 3, 4, 5]
print("\nAlready sorted:", arr4)
result4 = counting_sort(arr4.copy())
print("After Counting Sort:", result4)

# Test 5: Reverse sorted
arr5 = [5, 4, 3, 2, 1]
print("\nReverse sorted:", arr5)
result5 = counting_sort(arr5.copy())
print("After Counting Sort:", result5)

# Test 6: Single element
arr6 = [42]
print("\nSingle element:", arr6)
result6 = counting_sort(arr6.copy())
print("After Counting Sort:", result6)

print("\nComplexity Analysis:")
print("Time Complexity: O(n + k) where k is range of input")
print("Space Complexity: O(k)")
print("Best Case: O(n + k)")
print("Worst Case: O(n + k)")
print("Average Case: O(n + k)")
print("Stable: Yes")
print("Note: Non-comparison, efficient for small ranges")
